[core][telemetry/11] record histogram metric e2e

can-anyscale opened the pull request:
This is a series of PR to migrate metric collection from opencencus to openlemetry. For context on the existing components, see https://github.com/ray-project/ray/pull/53098.

--------

This PR adds support for (i) metric registration and (ii) metric value setting of `histogram` on the c++ side. 
It also supports histogram on the python side and glue everything together to make things work end-to-end.

**Note:**
Histogram metrics are exported as approximations in the new OpenTelemetry stack. This is because we attempt to reconstruct the original histogram data points—based on the internal representation sent from the C++ side—before exporting them to Prometheus. We evaluated several approaches:
- Solution 1: Approximate each bucket with its midpoint value. This is the approach we ultimately chose, as it offers a reasonable approximation of the histogram’s shape while keeping the implementation simple.
- Solution 2: Export a single data point per interval using sum / count. We rejected this option because it discards the original data point count, which can be important for interpretation.
- Solution 3: Export an additional metric on the C++ side to represent the sum per bucket for each histogram. We did not pursue this due to the added code complexity and challenges in atomically exporting two metrics within the same interval.
- Solution 4: Export each histogram data point as a separate gauge. This approach was dismissed due to potential performance overhead and increased implementation complexity.
- Solution 5: Implement a custom OpenTelemetry exporter to bypass the per-data-point recording model of the official Prometheus exporter. We chose not to go down this path to avoid adding significant complexity just for histogram metrics.

Test:
- CI
- I also pulled the histogram metric into a separated test

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/01] migrate python metric collection to opentelemetry

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/02] opentelemetry recorder grpc service

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/02] opentelemetry recorder grpc service

Signed-off-by: can <can@anyscale.com>

can-anyscale committed:
debugging (#53265)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

<!-- Please give a short summary of the change and the problem this
solves. -->

<!-- For example: "Closes #1234" -->

- [ ] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/05] refactor open_telemetry_metric_recorder.py

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/08] record counter metric e2e

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/08] record counter metric e2e

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/09] record sum metric e2e

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/09] record sum metric e2e

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/09] record sum metric e2e

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/09] record sum metric e2e

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/10] support custom gauge+counter+sum metrics

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][telemetry/11] support histogram metric on worker side

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

dayshah reviewed (commented):
why do we need to add this now if it's not used on the cpp side?

can-anyscale commented on src/ray/telemetry/open_telemetry_metric_recorder.h:
```diff
@@ -110,6 +115,8 @@ class OpenTelemetryMetricRecorder {
   // the recorder will only shutdown once.
   std::atomic<bool> is_shutdown_{false};
+  // The name of the meter used for this recorder.
+  const std::string meter_name_ = "ray";
```
reverting to `string` for now so the codes compile and i can test it e2e; will sync back with you before merging

can-anyscale committed:
[core][telemetry/12] record histogram metric e2e (#53927)

This is a series of PR to migrate metric collection from opencencus to
openlemetry. For context on the existing components, see
https://github.com/ray-project/ray/pull/53098.

------

This PR
- Support histogram metric e2e

Test:
- CI

---------

Signed-off-by: can <can@anyscale.com>
Signed-off-by: Cuong Nguyen <can@anyscale.com>

MengjinYan commented on src/ray/stats/tests/metric_with_open_telemetry_test.cc:
```diff
@@ -53,6 +53,19 @@ static ray::stats::Sum LegacyMetricSumTest("legacy_metric_sum_test",
                                            "",
                                            {"Tag1", "Tag2"});
 
+DECLARE_stats(metric_histogram_test);
```
Wondering should we add a test to verify the histogram metric?

MengjinYan commented on python/ray/tests/test_open_telemetry_metric_recorder.py:
```diff
@@ -2,7 +2,7 @@
 
 import pytest
-from opentelemetry.metrics import NoOpCounter, NoOpUpDownCounter
+from opentelemetry.metrics import NoOpCounter, NoOpUpDownCounter, NoOpHistogram
```
Wondering should we add a test for the histogram?

MengjinYan commented on src/ray/telemetry/open_telemetry_metric_recorder.cc:
```diff
@@ -165,6 +171,35 @@ void OpenTelemetryMetricRecorder::RegisterSumMetric(const std::string &name,
+  auto view = std::make_unique<opentelemetry::sdk::metrics::View>(
+      name,
+      description,
+      "",
```
Wondering should this be `meter_name_`?

MengjinYan commented on src/ray/telemetry/open_telemetry_metric_recorder.cc:
```diff
@@ -165,6 +171,35 @@ void OpenTelemetryMetricRecorder::RegisterSumMetric(const std::string &name,
+
+  auto instrument_selector =
+      std::make_unique<opentelemetry::sdk::metrics::InstrumentSelector>(
+          opentelemetry::sdk::metrics::InstrumentType::kHistogram, name, "");
```
```suggestion
          opentelemetry::sdk::metrics::InstrumentType::kHistogram, name, /*unit_filter=*/ "");
```

MengjinYan commented on src/ray/telemetry/open_telemetry_metric_recorder.cc:
```diff
@@ -165,6 +171,35 @@ void OpenTelemetryMetricRecorder::RegisterSumMetric(const std::string &name,
+      std::make_unique<opentelemetry::sdk::metrics::InstrumentSelector>(
+          opentelemetry::sdk::metrics::InstrumentType::kHistogram, name, "");
+  auto meter_selector =
+      std::make_unique<opentelemetry::sdk::metrics::MeterSelector>(meter_name_, "", "");
```
```suggestion
      std::make_unique<opentelemetry::sdk::metrics::MeterSelector>(meter_name_, /*meter_version=*/ "", /*schema_url=*/ "");
```

MengjinYan commented on src/ray/telemetry/open_telemetry_metric_recorder.cc:
```diff
@@ -165,6 +171,35 @@ void OpenTelemetryMetricRecorder::RegisterSumMetric(const std::string &name,
+      std::make_unique<opentelemetry::sdk::metrics::MeterSelector>(meter_name_, "", "");
+  meter_provider_->AddView(
+      std::move(instrument_selector), std::move(meter_selector), std::move(view));
+  auto instrument = GetMeter()->CreateDoubleHistogram(name, description, "");
```
```suggestion
  auto instrument = GetMeter()->CreateDoubleHistogram(name, description, /*unit=*/ "");
```

MengjinYan commented on python/ray/dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -562,6 +563,61 @@ async def ReportOCMetrics(self, request, context):
             logger.error(traceback.format_exc())
         return reporter_pb2.ReportOCMetricsReply()
 
+    def _export_histogram_data(
```
Seems that the `Export` function needs to be refactored to leverage the newly added functions

MengjinYan commented on src/ray/stats/stats.h:
Nice!!

MengjinYan commented on src/ray/telemetry/open_telemetry_metric_recorder.cc:
```diff
@@ -177,6 +183,41 @@ void OpenTelemetryMetricRecorder::RegisterSumMetric(const std::string &name,
+  auto aggregation_config =
+      std::make_shared<opentelemetry::sdk::metrics::HistogramAggregationConfig>();
+  aggregation_config->boundaries_ = buckets;
+  auto view = std::make_unique<opentelemetry::sdk::metrics::View>(
```
Found an example of creating the histogram view here: https://github.com/open-telemetry/opentelemetry-cpp/blob/main/examples/metrics_simple/metrics_ostream.cc#L93

Wondering whether we should follow the example to use the Factories to create the views or selectors?

can-anyscale replied:
nice, i'll take a look

MengjinYan replied:
Nit: Can we add it to the issue for the followups and link it here?

can-anyscale replied:
#resolved

MengjinYan commented on src/ray/stats/stats.h:
```diff
@@ -64,7 +64,6 @@ static inline void Init(
     int64_t max_grpc_payload_size = RayConfig::instance().agent_max_grpc_message_size()) {
   absl::MutexLock lock(&stats_mutex);
   if (StatsConfig::instance().IsInitialized()) {
-    RAY_CHECK(metrics_io_service_pool != nullptr);
```
Just curious, why this check is removed?

can-anyscale replied:
The open telemetry API doesn't need to use this service pool so I can remain null forever for that code path ;) (it only set to a non-null value in [1])

MengjinYan commented on python/ray/tests/test_metrics_agent.py:
```diff
@@ -713,6 +693,60 @@ def verify():
+        dashboard_export_addr,
+    ) = _setup_cluster_for_test
+
+    def test_cases():
```
Can we add tests for more data points fall into different buckets?

can-anyscale replied:
totally

can-anyscale commented:
@MengjinYan's comments - best effort to reconstruct data points for histogram metrics

MengjinYan commented on python/ray/dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -565,6 +566,125 @@ async def ReportOCMetrics(self, request, context):
+            bucket_values = []
+            total_estimated_sum = 0.0
+
+            for i, bucket_count in enumerate(data_point.bucket_counts):
```
Wondering is it possible to calculate the bucket average once during registration instead of every time during the metrics export?

can-anyscale replied:
#resolved

MengjinYan commented on python/ray/dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -565,6 +566,125 @@ async def ReportOCMetrics(self, request, context):
+                    (bucket_midpoint, bucket_count, lower_bound, upper_bound)
+                )
+
+            # Scale all values to preserve the actual sum while respecting boundaries
```
As discussed, as the sum will no matter be inaccurate, we should remove the logic to do the scale up and down.

MengjinYan commented on python/ray/dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -565,6 +566,125 @@ async def ReportOCMetrics(self, request, context):
+            bucket_values = []
+            total_estimated_sum = 0.0
+
+            for i, bucket_count in enumerate(data_point.bucket_counts):
```
Also, we should add a TODO here to update the document to communicate the inaccuracy of the sum of the histogram.

can-anyscale replied:
#resolved

MengjinYan reviewed (commented):
As discussed offline, probably we should document all the possible solutions and the reasoning why we pick the current solution in the description or a comment of this PR.

can-anyscale commented on python/ray/dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -576,41 +632,10 @@ async def Export(
-                                tag.key: tag.value.string_value
-                                for tag in data_point.attributes
-                            },
-                            # Note that all data points received from other Ray
```
#resolved

can-anyscale commented on python/ray/_private/telemetry/open_telemetry_metric_recorder.py:
```diff
@@ -99,6 +100,51 @@ def register_sum_metric(self, name: str, description: str) -> None:
+            # internal representation to approximated histogram data points.
+            for i in range(len(buckets)):
+                if i == 0:
+                    lower_bound = 0.0 if buckets[0] > 0 else buckets[0] * 0.5
```
ah nice catch!

can-anyscale commented:
@MengjinYan's comments

MengjinYan reviewed (approved):
Thanks!
